### ISSUE TITLE
MP rest client interface cannot be used as a CDI injected client and REST endpoint

### ISSUE THREAD

**Issue by pavolloffay** [quarkusio/quarkus#2148]
When I want to use a single interface to define REST interface and inject MP REST client as CDI bean I get a compilation error.

```java
@Path("/")
@RegisterRestClient
public interface GreetingService {

  @GET
  @Path("/hello")
  @Produces(MediaType.TEXT_PLAIN)
  String hello();

  @GET
  @Path("/bonjour")
  @Produces(MediaType.TEXT_PLAIN)
  String bonjour();

  @GET
  @Path("/conversation")
  @Produces(MediaType.TEXT_PLAIN)
  String conversation();
}
```

```java
@Path("/")
public class GreetingResource implements GreetingService {

    @Inject
    @RestClient
    private GreetingService greetingService;

    @Override
    public String hello() {
        return "hello";
    }

    @Override
    public String bonjour() {
        return "bonjour";
    }

    @Override
    public String conversation() {
        return greetingService.hello() + " -> " + greetingService.bonjour();
    }
}
```

```
./mvnw compile quarkus:dev                                                                                                                                                           1:57 
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------< org.acme:tracing-example >----------------------
[INFO] Building tracing-example 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ tracing-example ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ tracing-example ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- quarkus-maven-plugin:0.13.3:dev (default-cli) @ tracing-example ---
Listening for transport dt_socket at address: 5005
13:57:05,504 INFO  [io.qua.dep.QuarkusAugmentor] Beginning quarkus augmentation
13:57:05,649 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Alpha4
13:57:05,917 ERROR [io.qua.dev.DevModeMain] Failed to start quarkus: java.lang.RuntimeException: org.jboss.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.quickstart.GreetingService and qualifiers [@RestClient]
	- java member: org.acme.quickstart.GreetingResource#greetingService
	- declared on CLASS bean [types=[org.acme.quickstart.GreetingService, org.acme.quickstart.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.quickstart.GreetingResource]
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:134)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:108)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: org.jboss.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.quickstart.GreetingService and qualifiers [@RestClient]
	- java member: org.acme.quickstart.GreetingResource#greetingService
	- declared on CLASS bean [types=[org.acme.quickstart.GreetingService, org.acme.quickstart.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.quickstart.GreetingResource]
	at org.jboss.builder.Execution.run(Execution.java:123)
	at org.jboss.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:136)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:110)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:99)
	... 2 more
Caused by: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.quickstart.GreetingService and qualifiers [@RestClient]
	- java member: org.acme.quickstart.GreetingResource#greetingService
	- declared on CLASS bean [types=[org.acme.quickstart.GreetingService, org.acme.quickstart.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.quickstart.GreetingResource]
	at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:739)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:276)
	at io.quarkus.arc.processor.BeanProcessor.process(BeanProcessor.java:153)
	at io.quarkus.arc.deployment.ArcAnnotationProcessor.build(ArcAnnotationProcessor.java:259)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:506)
	at org.jboss.builder.BuildContext.run(BuildContext.java:413)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1416)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.quickstart.GreetingService and qualifiers [@RestClient]
	- java member: org.acme.quickstart.GreetingResource#greetingService
	- declared on CLASS bean [types=[org.acme.quickstart.GreetingService, org.acme.quickstart.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.quickstart.GreetingResource]
	at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:326)
	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:375)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:268)
	... 14 more

```

**Comment by kenfinnigan** [comment]
Is it a valid use case to have a service and client with the same endpoint definitions in a single application?

**Comment by pavolloffay** [comment]
I don't know, if it's not prohibited by rest client spec I guess anybody can use it. @pilhuhn used to demo this scenario to showcase how easy is to get a client for a rest service.

**Comment by kenfinnigan** [comment]
REST Client spec has always been in favor of re-using an interface between client and server, but not in the same application/microservice

**Comment by pavolloffay** [comment]
If there is no technical blocker I don't see a reason why it shouldn't be allowed.

**Comment by gsmet** [comment]
Not saying it's something we want but...

We could probably make it work in the case where we have a `@RegisterRestClient` annotation. This one is easy.

And it would maybe be possible to make it work in the other case by collecting the types marked with `@RestClient` and create REST clients for them. Maybe it should be our approach instead of collecting classes with `@Path` and no implementations.

WDYT @kenfinnigan ?

**Comment by kenfinnigan** [comment]
The issue is we have special handling for service vs client code.

And we need to be able to distinguish between an interface for client and one for service. If an interface used for a client is also implemented for a service, there's no way to know it's for a client too

**Comment by kenfinnigan** [comment]
The issue is that `@RegisterRestClient` and `@RestClient` are purely for CDI use cases.

There would be no way to know, without trawling through application code, that an interface is used as a client with the `RestClientBuilder` in code

**Comment by pavolloffay** [comment]
@kenfinnigan feel free to close or proceed. I don't know at this point what the impl for this would look like. Maybe we just need better docs somewhere.

When the same interface is used for a service and a client there is also another exception:

```
2019-04-18 15:34:26,955 ERROR [io.und.request] (XNIO-1 task-1) UT005023: Exception handling request to /conversation: org.jboss.resteasy.spi.UnhandledException: com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface org.acme.quickstart.GreetingService, interface org.jboss.resteasy.client.jaxrs.internal.proxy.ResteasyClientProxy] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:234)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:72)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:175)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:43)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:52)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentTemplate$7$1$1.call(UndertowDeploymentTemplate.java:416)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1382)
	at java.lang.Thread.run(Thread.java:748)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:473)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface org.acme.quickstart.GreetingService, interface org.jboss.resteasy.client.jaxrs.internal.proxy.ResteasyClientProxy] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
	at com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:102)
	at com.oracle.svm.reflect.proxy.DynamicProxySupport.getProxyClass(DynamicProxySupport.java:112)
	at java.lang.reflect.Proxy.getProxyClass0(Proxy.java:50)
	at java.lang.reflect.Proxy.newProxyInstance(Proxy.java:719)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ProxyBuilderImpl.build(ProxyBuilderImpl.java:169)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ProxyBuilderImpl.build(ProxyBuilderImpl.java:108)
	at io.quarkus.smallrye.restclient.runtime.RestClientBuilderImpl.build(RestClientBuilderImpl.java:205)
	at org.acme.quickstart.GreetingResource.conversation(GreetingResource.java:28)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:151)
	at org.jboss.resteasy.core.MethodInjectorImpl.lambda$invoke$3(MethodInjectorImpl.java:122)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:110)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:122)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:568)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:442)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:396)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:398)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:367)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invoke$1(ResourceMethodInvoker.java:341)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:341)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:477)
	... 51 more
```

**Comment by gsmet** [comment]
Yes, we wouldn't cover the RestClientBuilder usage. TBH, I really would like the REST client interfaces to have a mandatory annotation. It's cute to not requiring it but it makes things harder for Quarkus.

I still think that it would be better to include them forcibly in the discovery (them being the classes with `@RegisterRestClient` and the types injected with `@RestClient`).

**Comment by kenfinnigan** [comment]
We can always add an annotation to SmallRye REST Client to indicate an interface is used as a client, to get around the problem.

Or just support `@RegisterRestClient` for that purpose. Not sure if there's any issues with doing that

**Comment by pilhuhn** [comment]
What I did in the past is here in the video https://youtu.be/sL_n85qrppI?t=86 - basically I create an interface in the backend, then copy that source file to my frontend and decorate it with `@RegisterRestClient` And then at minute 2:04 I use that client interface via `@Inject @RestClient` .

I think the "effort" to put `@RegisterRestClient` on the interface can be neglected.

**Comment by kenfinnigan** [comment]
@pilhuhn, do you mean not require `@RegisterRestClient`? At the moment it's used to indicate an interface should have a CDI Bean created for it

**Comment by pilhuhn** [comment]
Ah no :)  My last sentence meant that it is not a huge deal if a developer has to put that annotation on the interface source file that they have received from the backend developer.

**Comment by kenfinnigan** [comment]
gotcha

**Comment by kenfinnigan** [comment]
I think for now we will close this.

If it becomes a real use problem we can re-evaluate

**Comment by fterella** [comment]
Hi,

I'm using quarkus in order to build several microservices and I'm struggling with this scenario. My idea is to create a custom api gateway that calls several microservices. I've defined the interfaces as rest client and signature for the effective rest endpoints, I've implemented them (on different projects) as endpoints, and finally I'm trying to develop a gateway that uses the interfaces both as signature and rest client in order to call the right microservice.
I know is a little bit complicated to describe but, I have:
- an api gateway that exposes endpoints, verifies some simple properties (grants and so on) and calls the right microservice. This gateway implements the interfaces and uses them as injected rest client. After the checks, this layer basically acts as a simple proxy
- several microservices that implements the interfaces with custom business logic. This layer is the responsible for the business logic and it's the layer called by the gateway
- several interfaces that I'm using as client definition (with @RegisterRestClient) and signature definition for the previous described components

In this scenario, inside my gateway I'm not enabled to implement the interface (in order to follow the signature) and in the same time inject it as rest client.

Is it possible to consider this scenario again?

**Comment by kenfinnigan** [comment]
I believe this scenario can be done by not including `@RegisterRestClient` on the parent interface that contains all the methods, and then create a child interface that just has that annotation to use for REST Client.

**Comment by fterella** [comment]
I've tried and it is not sufficient to create a simple child interface with @RegisterRestClient annotation, because in that scenario a similar error occurs (during the maven install goal). In order to build and execute the gateway, I must also include again the @Path annotation into the child interface. With this modification I was able to run the gateway as expected.
Thank you for the tip!

**Comment by gsmet** [comment]
@kenfinnigan we had several complaints about this not being possible. Any chance we could come up with a plan to fix this use case?

**Comment by leonidr** [comment]
@kenfinnigan  - debugging the RestClientProcessor i've found that the use case we ask for is blocked by isRestClientInterface method checking that there are no implementations of the interface  ==> "index.getAllKnownImplementors(classInfo.name()).isEmpty()"
I've seen that REST endpoint implementing such interface is registered with @Typed annotation scoped to endpoint implementing class only - e.g. it will not be found when injecting the interface.
What is the limitation currently that prevents us from unlocking the possibility?

**Comment by kenfinnigan** [comment]
I believe it's a technical effort to investigate all the use cases and come up with a way that works for them all.

Please open a new issue if you would like it re-evaluated
